src/crystal/system/unix/io_uring.cr

```diff
@@ -0,0 +1,617 @@
+    end
+
+    # The submission queue just stores indices into the `submission_entries` array.
+    sqe = @submission_entries + index
+
+    # This submission queue entry is reused! Clean it up.
+    Intrinsics.memset(sqe, 0_u8, sizeof(Syscall::IoUringSqe), false)
+
+    # Populate fields and submit the index into the submission queue.
+    yield sqe
+    sqe.value.user_data = user_data_pointer.address
+    sqe.value.opcode = opcode
+
+    # If there is a timeout, we submit a LINK_TIMEOUT as well. It will make the
+    # original event fail with ECANCELED if it doesn't complete in time.
+    unless timeval.null?
+      until (index_timeout = @submission_queue.consume_free_index) != UInt32::MAX
+        process_completion_events(blocking: false)
+      end
+
+      sqe_timeout = @submission_entries + index_timeout
+      Intrinsics.memset(sqe_timeout, 0_u8, sizeof(Syscall::IoUringSqe), false)
+      sqe_timeout.value.opcode = Syscall::IoUringOp::LINK_TIMEOUT
+      sqe_timeout.value.addr = timeval.address
+      sqe_timeout.value.len = 1u32
+      sqe.value.flags = sqe.value.flags | Syscall::IoUringSqeFlags::IO_LINK
+
+      @submission_queue.push(index)
+      @submission_queue.push(index_timeout)
+    else
+      @submission_queue.push(index)
+    end
+  end
+
+  # Safer version of `submit!`, this will block the current fiber until the request completes
+  # and will then return the completion result.
+  private def submit(opcode : Syscall::IoUringOp, *, timeout : ::Time::Span? = nil)
+    user_data = UserData.new
+    timeval = make_timeval(timeout || ::Time::Span::ZERO)
+    timeval_ptr = timeout ? pointerof(timeval) : Pointer(LibC::Timeval).null
+    submit!(opcode, pointerof(user_data), timeval: timeval_ptr) { |sqe| yield sqe }
+    sleep # The completion event will wake up this fiber.
+    user_data.res
+  end
+
+  # Enters the kernel to process events. This can be called either by the event loop when no fiber has
+  # work to do (most likely all of them are waiting for completion events) or when the submission queue
+  # is full and we are trying to submit more requests. The `blocking` parameter dictates if it should
+  # block until at least one completion entry is available or if it should return as soon as possible.
+  def process_completion_events(*, blocking = true)
+    # Consume available completion events and enqueue their fibers for execution.
+    completed_some = false
+    @completion_queue.consume_all do |cqe|
+      completed_some = true
+
+      # cqe.user_data is zero when this is the completion for a LINK_TIMEOUT event.
+      unless cqe.user_data == 0
+        # Store result and enqueue the fiber that is waiting for this completion.
+        user_data_ptr = Pointer(UserData).new(cqe.user_data)
+        user_data_ptr.value.res = cqe.res
+        user_data_ptr.value.fiber.enqueue
+      end
+    end
+
+    # If we consumed at least one event, then there are fibers with work to do.
```

Comment:
I wonder if it is possible to get only the link_timeout back without the other one at the same time. Relevant question because the order you get them back was switched in some kernel version.

Reply:
I actually don't do any action with the link_timeout response, it is just ignored. I actually handle the fact that the original event fails with ECANCELED if the timeout expires. So the order doesn't matter.

Reply:
This is only true because the IORING_OP_ASYNC_CANCEL operation won't be needed and doesn't need to be supported.

Reply:
The point I was making is that completed_some is set to true unconditionally so the following line could potentially return without having any available stuff to do. But perhaps it doesn't matter.

Reply:
Fixed.